rewindable-index/src/RewindableIndex/Storable.hs

```diff
@@ -0,0 +1,358 @@
+   data and whenever we have a rollback restore the version which is the target of the
+   rollback.
+
+   This is what the next class is meant to do.
+-}
+class Rewindable h m where
+  rewindStorage :: StorablePoint h -> h -> m (Maybe h)
+
+{-
+   Another feature of indexers is the ability to resume from a previously stored event.
+   One way of implementing this is to make sure that there is always at least one event
+   present on disk, event which includes the slot number when it was generated.
+
+   This function should return the most recent resume point. The plan is to support
+   multiple resume points, but for now there is no way to ensure that no data duplication
+   happens (the same events get reinserted).
+
+   This will not be a problem if the indexer learns how to remove the `old` slots
+   so there will be no duplication. This may be implemented later if the API users
+   request it.
+-}
+class Resumable h m where
+  resumeFromStorage :: h -> m [StorablePoint h]
+
+{-
+   The next class is witnessing the fact that events contain enough information to
+   retrieve the point when they were produced.
+-}
+class HasPoint e p where
+  getPoint :: e -> p
+
+{-
+   The configuration includes hints about the amount of events stored in memory and
+   on disk. This information is used by the storage engine to decide how much memory
+   to allocated and when to flush the memory buffer or roll the disk events into
+   an aggregate database structure.
+-}
+data Config = Config
+  { _memoryBufferSize :: Int
+  , _diskBufferSize   :: Int
+  } deriving (Show, Eq)
+$(Lens.makeLenses ''Config)
+
+data Storage h m = Storage
+  { _events :: VM.MVector (PrimState m) (StorableEvent h)
+  , _cursor :: Int
+  }
+$(Lens.makeLenses ''Storage)
+
+data State h m = State
+  { _config  :: Config
+  , _storage :: Storage h m
+  , _handle  :: h
+  }
+$(Lens.makeLenses ''State)
+
+emptyState
+  :: PrimMonad m
+  => Int
+  -> Int
+  -> h
+  -> m (State h m)
+emptyState memBuf dskBuf hdl = do
+  v <- VM.new memBuf
+  pure $ State { _config = Config { _memoryBufferSize = memBuf
+                                  , _diskBufferSize   = dskBuf
+                                  }
+               , _storage = Storage { _events = v
+                                    , _cursor = 0
+                                    }
+               , _handle = hdl
+               }
+
+-- Get events from the memory buffer.
+getMemoryEvents
+  :: Storage h m
+  -> V.MVector (PrimState m) (StorableEvent h)
+getMemoryEvents s = VM.slice 0 (s ^. cursor) (s ^. events)
+
+-- Get events from memory buffer and disk buffer.
+getEvents
+  :: Buffered h m
+  => PrimMonad m
+  => State h m
+  -> m [StorableEvent h]
+getEvents s = do
+  memoryEs <- getMemoryEvents (s ^. storage)
+              & V.freeze <&> V.toList
+  diskEs   <- getStoredEvents (s ^. handle)
+  pure $ diskEs ++ memoryEs
+
+insert
+  :: Buffered h m
+  => PrimMonad m
+  => StorableEvent h
+  -> State h m
+  -> m (State h m)
```

Comment:
Did you consider using a stack monad to handle `Config` and `State`? It looks similar to `MonadReader` and `MonadState` if we decompose them and it may simplify the logic.
I even wonder if in the end, `h` implementation shouldn't be hidden in the definition of `m` and then the ground type of your family definition would be the event.

Reply:
I do intent to use something like `mtl` for the API once it gets stable enough.